### DISCUSSION TITLE
Does Timeshape resolve country code as well?

### DISCUSSION THREAD

**Discussion by akefirad** [RomanIakovlev/timeshape#148] [Q&A]
Hi there,
We've got a very simple use case; we'd like to resolve country-code and timezone by lat/lon. Theoretically Timezone IDs are tightly coupled with country codes. So I'm hoping Timeshape supports both. Is that right?
If it doesn't support it, is there any chance to support it in future?
We're happy to contribute if the feature gets prioritized.
Thanks.

**Comment by RomanIakovlev** [comment]
Hi! Thanks for your interest in contributing to Timeshape, it's much appreciated.

Doing a direct lookup from coordinate to country would probably at least double the size of Timeshape JAR, because we'll have to store the country-level geometry in addition to time zone-level geometry. Timeshape JAR is already quite large at 22MB, so I'd be hesitant to increase it further.

You can try and do a lookup from time zone to country, but I'm not sure it's possible to do it reliably. Even if it is possible, this feature doesn't have to be a part of Timeshape, as it would simply be a static mapping from `java.time.ZoneId` to a `String`, presuming it's a one to one mapping, which I'm not sure about.

All in all, I'm not 100% sure there's a nice way to add this feature to Timeshape. If you could make a more detailed suggestion, I'd have more information to think about.

**Comment by akefirad** [comment]
Thanks for the reply.

Right. Resolving country code using TZ is not reliable.

The point about the size is also fair. What if it’s a separate artifact: one with both TZ and Country Code?

**Comment by RomanIakovlev** [comment]
I think a separate artefact would work, but I'd lean towards having it containing just the countries, not both countries and timezones. This would give us more flexibility with releasing the new versions, since we'd have to release the combined artefact as soon as either time zone or country information changes. It will also open an entirely new use case for users who only need to do country lookup.

As for users who need both time zones and countries, using both Timeshape and this hypothetical new country-only artefact simultaneously should work just fine. There will be no performance hit, because two point-in-polygon lookups will be needed anyway, one for time zone, one for country.

P.S. If I'd be interested in this feature myself, I'd still explore creating the `ZoneId` -> `Country` mapping separately of Timeshape. I have a hunch it should be possible to construct a 'good enough' mapping here, especially if you're only interested in some parts of the world. 

If you look at the list of all known time zones from which Timeshape is built [here](https://github.com/evansiroky/timezone-boundary-builder/releases/download/2024a/timezone-names.json), you'll notice that there's only 58 zone ids starting with e.g. `Europe`. So if you're only interested in Europe, it's possible to hand-craft a mapping between 58 entries and their matching countries with very reasonable amount of effort. For `Africa` it's only 52 zones as well. It's more than that for `America`, but, again, not an insurmountable amount of work, especially if you ask ChatGPT to do it for you. :)

**Reply by RomanIakovlev** [reply]
P.P.S. FWIW I've asked Claude AI to create this time zone to country mappings, it took only a couple of minutes. Maybe it will be useful for you, so here it goes.
[timezone-country-mapping.zip](https://github.com/user-attachments/files/16863307/timezone-country-mapping.zip)

Obviously validate the results before using it yourself, since it came from AI, which are notoriously unreliable.

I've skipped Australia time zones, because they all belong to the same country, as well as the `Etc/GMT*` time zones, since they don't belong to any country.

**Comment by akefirad** [comment]
Hi again, sorry for the delay, coming back to this. I saw in the README.md that you're working on [GeoBundle](https://geobundle.com). Is there any progress there? We're happy to become a paid customer if there's a clear timeline. Thanks.

**Reply by RomanIakovlev** [reply]
Hi, and thanks for your interest in GeoBundle. It's been on a back burner for a while, and I can't offer you a clear timeline there. However we can discuss some alternative engagements with you on this topic. Feel free to email me with your contact details, and we can take it from there.

**Reply by akefirad** [reply]
Sure, sent.

**Reply by akefirad** [reply]
Hi @RomanIakovlev , did you get my email? Looking forward to hearing from you.
